external-crates/move/move-bytecode-verifier/src/abstract_stack.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines an abstract stack used by various verifier passes
+
+use std::cmp::Ordering;
+use std::fmt;
+
+#[derive(Default)]
+/// An abstract value that compresses runs of the same value to reduce space usage
+pub struct AbsStack<T> {
+    values: Vec<(u64, T)>,
+}
+
+impl<T: Eq + Clone> AbsStack<T> {
+    /// Creates an empty stack
+    pub fn new() -> Self {
+        Self { values: vec![] }
+    }
+
+    /// Returns true iff the stack is empty
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    /// Pushes a single value on the stack
+    pub fn push(&mut self, item: T) {
+        self.push_n(item, 1)
+    }
+
+    /// Pops a single value off the stack, erroring if empty
+    pub fn pop(&mut self) -> Result<T, AbsStackError> {
+        let Some((count, last)) = self.values.last_mut() else {
+            return Err(AbsStackError::EmptyStack)
+        };
+        debug_assert!(*count > 0);
+        Ok(if *count <= 1 {
+            let (_, last) = self.values.pop().unwrap();
+            last
+        } else {
+            *count -= 1;
+            last.clone()
+        })
+    }
+
+    /// Push n copies of an item on the stack
+    pub fn push_n(&mut self, item: T, n: u64) {
+        if n == 0 {
+            return;
+        }
+
+        match self.values.last_mut() {
+            Some((count, last_item)) if &item == last_item => {
+                debug_assert!(*count > 0);
+                *count += n
```

Comment:
Accounting for the true overflow here seems like some worthwhile hardening. 

I'm not sure about the logical case though, especially since we don't expose a size/len function.